server-common/src/main/java/org/apache/kafka/server/group/share/DefaultStatePersister.java

```diff
@@ -32,94 +40,156 @@
+                            if (!partMap.containsKey(partitionData.partition())) {
+                                partMap.put(partitionData.partition(), new CompletableFuture<>());
+                            }
+                            return stateManager.new ReadStateHandler(
+                                    groupId,
+                                    topicData.topicId(),
+                                    partitionData.partition(),
+                                    partitionData.leaderEpoch(),
+                                    partMap.get(partitionData.partition()));
+                        })
+                        .collect(Collectors.toList()))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        for (PersisterStateManager.PersisterStateManagerHandler handler : handlers) {
+            stateManager.enqueue(handler);
+        }
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+                .flatMap(map -> map.values().stream()).toArray(CompletableFuture[]::new));
+
+        // Transform the combined CompletableFuture<Void> into CompletableFuture<ReadShareGroupStateResult>
+        return combinedFuture.thenApply(v -> {
+            List<TopicData<PartitionAllData>> topicsData = futureMap.entrySet().stream()
+                    .map(topicId -> {
+                        List<PartitionAllData> partitionAllData = futureMap.get(topicId).values().stream()
+                                .map(future -> {
+                                    try {
+                                        ReadShareGroupStateResponse partitionResponse = future.get();
+                                        return partitionResponse.data().results().get(0).partitions().stream()
+                                                .map(partitionResult -> PartitionFactory.newPartitionAllData(
+                                                        partitionResult.partition(),
+                                                        partitionResult.stateEpoch(),
+                                                        partitionResult.startOffset(),
+                                                        partitionResult.errorCode(),
+                                                        partitionResult.errorMessage(),
+                                                        partitionResult.stateBatches().stream().map(PersisterStateBatch::from).collect(Collectors.toList())
+                                                ))
+                                                .collect(Collectors.toList());
+                                    } catch (InterruptedException | ExecutionException e) {
+                                        throw new RuntimeException(e);
```

Comment:
I wonder whether these errors actually ought somehow to be populating the partitionResponse. Otherwise, I think you'll get gaps in the response.

Reply:
Made the changes in the latest commit